snuba/query/parser/conditions.py

```diff
@@ -0,0 +1,225 @@
+from typing import Any, Callable, Optional, OrderedDict, Sequence, TypeVar
+
+from snuba.datasets.dataset import Dataset
+from snuba.query.expressions import (
+    Argument,
+    Column,
+    Expression,
+    Lambda,
+    Literal,
+    FunctionCall,
+)
+from snuba.query.conditions import (
+    binary_condition,
+    BooleanFunctions,
+    OPERATOR_TO_FUNCTION,
+)
+from snuba.query.parser.functions import parse_function_to_expr
+from snuba.query.schema import POSITIVE_OPERATORS
+from snuba.util import is_condition, is_function, QUOTED_LITERAL_RE
+
+
+TExpression = TypeVar("TExpression")
+
+
+class InvalidConditionException(Exception):
+    pass
+
+
+def parse_conditions(
+    operand_builder: Callable[[Any], TExpression],
+    and_builder: Callable[[Sequence[TExpression]], Optional[TExpression]],
+    or_builder: Callable[[Sequence[TExpression]], Optional[TExpression]],
+    unpack_array_condition_builder: Callable[[TExpression, str, Any], TExpression],
+    simple_condition_builder: Callable[[TExpression, str, Any], TExpression],
+    dataset: Dataset,
+    conditions: Any,
+    array_join: Optional[str],
+    depth: int = 0,
+) -> Optional[TExpression]:
+    """
+    Return a boolean expression suitable for putting in the WHERE clause of the
+    query.  The expression is constructed by ANDing groups of OR expressions.
+    Expansion of columns is handled, as is replacement of columns with aliases,
+    if the column has already been expanded and aliased elsewhere.
+
+    operand_builder: Builds the TExpression representing the left hand side
+      of a simple condition. This can be as nested as the user wants
+    and_builder / or_builder: Combine a list of expressions in AND/OR
+    unpack_array_condition_builder: Deals with a special case where we unpack conditions
+      on array columns. More details in the code.
+    simple_condition_builder: Generates a simple condition made by expression on the
+      left hand side, an operator and a literal on the right hand side.
+    """
+    from snuba.clickhouse.columns import Array
+
+    if not conditions:
+        return None
+
+    if depth == 0:
+        # dedupe conditions at top level, but keep them in order
+        sub = OrderedDict(
```

Comment:
```
snuba/query/parser/conditions.py:61: error: "TypeAlias" not callable
```

This is now new, but I think this is indicating that this should be using `collections.OrderedDict` instead of the generic variant.